packages/client/src/client/Web/serviceWorkerRegistration.ts

```diff
@@ -53,51 +56,78 @@ export function register(config?: Config) {
         })
       } else {
         // Is not localhost. Just register service worker
-        registerValidSW(swUrl, config)
+        registerServiceWorker(swUrl, config)
       }
     })
   }
 }
 
-function registerValidSW(swUrl: string, config?: Config) {
+function addStateChangeHandler(
+  serviceWorkerInstance: ServiceWorker,
+  registration: ServiceWorkerRegistration,
+  registrationConfig: Config | undefined,
+) {
+  serviceWorkerInstance.onstatechange = () => {
+    console.debug(
+      "'installing' SW changed state to: ",
+      serviceWorkerInstance.state,
+    )
+    if (serviceWorkerInstance.state === "installed") {
+      if (navigator.serviceWorker.controller) {
+        // At this point, the updated precached content has been fetched,
+        // but the previous service worker will still serve the older
+        // content until all client tabs are closed.
+        console.debug(
+          "New content is available and will be used when all " +
+            "tabs for this page are closed. See https://cra.link/PWA.",
+        )
+        serviceWorkerInstance.postMessage({ type: "SKIP_WAITING" })
+      } else {
+        // At this point, everything has been precached.
+        // It's the perfect time to display a
+        // "Content is cached for offline use." message.
+        console.debug("Content is cached for offline use.")
+
+        // Execute callback
+        if (registrationConfig && registrationConfig.onSuccess) {
+          registrationConfig.onSuccess(registration)
+        }
+      }
+    } else if (serviceWorkerInstance.state === "activated") {
+      if (registrationConfig && registrationConfig.onUpdate) {
+        registrationConfig.onUpdate(registration)
+      }
+    }
+  }
+}
+
+function registerServiceWorker(swUrl: string, config?: Config) {
   navigator.serviceWorker
     .register(swUrl)
     .then((registration) => {
-      registration.onupdatefound = () => {
+      console.debug("SW Reg: handling ..", registration)
+
+      if (registration.waiting) {
+        const waitingWorker = registration.waiting
+        console.debug(
+          "'waiting' SW found .. skip waiting in case",
+          waitingWorker,
+        )
+        waitingWorker.postMessage({ type: "SKIP_WAITING" })
+        addStateChangeHandler(waitingWorker, registration, config)
+      }
+
+      registration.addEventListener("updatefound", (event) => {
```

Comment:
on register, the usual case will be that a new SW is found remotely, and this is pulled down with its friends, the new hash-coded JS modules and when that is all done this "installing" SW will transition to "installed" .. this doesn't mean it is actively running on (and controlling) the page - that's what "activated" is for